### PR TITLE
Frc issues

### PR DESCRIPTION
**Pull request by bassrock**
This adds support beginUpdating and endUpdating like a regular UITableView. This is needed because FRC tends to insert multiple objects out of order.

This also adds support for the AppExtensionAPI which is needed for WatchKit Apps.

### CODE REVIEW

**Comment by rnystrom** [comment]
Awesome, thanks for adding this! Could you do me a favor and update it to use 2 spaces instead of 4 so there isn't so much stuff that's changed? That way it will be a little easier to read the changes.

Thanks!

**Comment by bassrock** [comment]
OK got it. Had to change the setting in Xcode to get it to work and save properly.

File: IGInterfaceDataTable.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -240,6 +240,7 @@
 		29BE4ED81A97A59F0025C38E /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				APPLICATION_EXTENSION_API_ONLY = YES;
```
Comment (rnystrom) [inline IGInterfaceDataTable.xcodeproj/project.pbxproj]: How about we remove this and put up another PR? Just so we can keep PRs localized to a specific change in case we need to revisit a change down the road.

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.h [hunk 2]
```diff
@@ -242,4 +259,4 @@
  */
 - (void)tableDidSelectFooter:(WKInterfaceTable *)table;
 
-@end
+@end
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.h]: Let's keep the newlines please.

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 3]
```diff
@@ -472,4 +545,4 @@ - (void)tableDidSelectFooter:(WKInterfaceTable *)table {
   // for subclassing
 }
 
-@end
+@end
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: Same note re: newlines

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.h [hunk 4]
```diff
@@ -178,6 +178,23 @@
...
+
+/**
+ * Stops the updates to the table to allow normal inserts
+ */
+- (void)endUpdates;
+
+/**
+ * Update Table Row
+ * @param indexPath The index path of the row to update
+ * @discussion This will only insert section @a headers and not rows. To insert rows see
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.h]: I think the `@discussion` needs finishing!

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.h [hunk 5]
```diff
@@ -178,6 +178,23 @@
...
+ */
+- (void)beginUpdates;
+
+/**
+ * Stops the updates to the table to allow normal inserts
+ */
+- (void)endUpdates;
+
+/**
+ * Update Table Row
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.h]: Let's change to:
>Update a row in the table for a given index path.

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 6]
```diff
@@ -64,6 +64,32 @@ - (NSString *)description {
 
 @implementation WKInterfaceTable (IGInterfaceDataTable)
 
+- (BOOL)updating {
+  NSNumber *number = objc_getAssociatedObject(self, @selector(updating));
+  return [number boolValue];
+}
+
+- (void)setUpdating:(BOOL)updating {
+  NSNumber *number = [NSNumber numberWithBool: updating];
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: super nit-picky, can we remove the whitespace before the `updating` parameter?

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 7]
```diff
@@ -64,6 +64,32 @@ - (NSString *)description {
 
 @implementation WKInterfaceTable (IGInterfaceDataTable)
 
+- (BOOL)updating {
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: How about making the getter `isUpdating`?

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 8]
```diff
@@ -341,6 +381,39 @@ - (void)insertRowsAtIndexPaths:(NSArray *)indexPaths withRowType:(NSString *)row
...
 
+- (void)beginUpdates {
+  self.updating = true;
+  self.rowIndexesToAdd = [[NSMutableArray alloc] init];
+  self.rowTypesToAdd = [[NSMutableDictionary alloc] init];
+}
+
+- (void)endUpdates {
+  if(!self.updating) {
+    return;
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: Let's also add an assert here above the if-statement so that if you're debugging you can catch that begin/end updates aren't being called in sync:
```
NSAssert(self.updating, @"Calling -endUpdates without first calling -beginUpdates.");
```

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 9]
```diff
@@ -341,6 +381,39 @@ - (void)insertRowsAtIndexPaths:(NSArray *)indexPaths withRowType:(NSString *)row
...
+  }
+
+  [self.rowIndexesToAdd sortUsingComparator:^NSComparisonResult(id obj1, id obj2) {
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: Let's strictly cast both parameters to `NSIndexPath`

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 10]
```diff
@@ -341,6 +381,39 @@ - (void)insertRowsAtIndexPaths:(NSArray *)indexPaths withRowType:(NSString *)row
...
+    NSInteger r1 = [obj1 row];
+    NSInteger r2 = [obj2 row];
+    if (r1 > r2) {
+      return (NSComparisonResult)NSOrderedDescending;
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: Is this cast required?

File: IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m [hunk 11]
```diff
@@ -341,6 +381,39 @@ - (void)insertRowsAtIndexPaths:(NSArray *)indexPaths withRowType:(NSString *)row
...
+    if (r1 < r2) {
+      return (NSComparisonResult)NSOrderedAscending;
+    }
+    return (NSComparisonResult)NSOrderedSame;
+  }];
+
+  self.updating = false;
+
+  [self.rowIndexesToAdd enumerateObjectsUsingBlock:^(NSIndexPath *indexPath, NSUInteger idx, BOOL *stop) {
+    [self insertRowsAtIndexPaths:@[indexPath] withRowType:[self.rowTypesToAdd objectForKey:indexPath]];
```
Comment (rnystrom) [inline IGInterfaceDataTable/WKInterfaceTable+IGInterfaceDataTable.m]: We're using subscripting throughout the project, let's keep it consistent here: `self.rowTypesToAdd[indexPath]`

**Comment by rnystrom** [comment]
Looking really good! I like the change. Don't worry about all of my notes, most of it is stylistic so we keep everything consistent w/ the current project.

Also just a note, when we're done and ready to merge we should squash all of the commits into one and then it'll be good. I can help you w/ that too when we get there.

Great work!

**Comment by rnystrom** [comment]
Great stuff! Last part so we can clean up the commit history. Just do an interactive rebase and squash all of the commits into one:

- `git reset -soft HEAD~6`
- `git commit`
- Add a commit message for this PR
- `git push -f origin/FRCIssues`

That should squish the last 6 commits into 1 so we have a nice clean history. Once that's up we're good to merge!

**Comment by bassrock** [comment]
@rnystrom ok should be good

**Comment by rnystrom** [comment]
Hell ya! This is awesome. I'm going to wrap up some stuff w/ unit tests working now that Travis is humming along, and we'll bump the version and announce!
